frappe/rate_limiter.py

```diff
@@ -82,37 +82,50 @@ def respond(self):
 		if self.rejected:
 			return Response(_("Too Many Requests"), status=429)
 
-def rate_limit(key: str, limit: Union[int, Callable] = 5, seconds: int= 24*60*60, methods: Union[str, list]='ALL'):
+def rate_limit(key: str = None, limit: Union[int, Callable] = 5, seconds: int = 24*60*60, methods: Union[str, list] = 'ALL', ip_based: bool = True):
 	"""Decorator to rate limit an endpoint.
 
 	This will limit Number of requests per endpoint to `limit` within `seconds`.
 	Uses redis cache to track request counts.
 
-	:param key: Key is used to identify the requests uniqueness
+	:param key: Key is used to identify the requests uniqueness (Optional)
 	:param limit: Maximum number of requests to allow with in window time
 	:type limit: Callable or Integer
 	:param seconds: window time to allow requests
 	:param methods: Limit the validation for these methods.
 		`ALL` is a wildcard that applies rate limit on all methods.
 	:type methods: string or list or tuple
+	:param ip_based: flag to allow ip based rate-limiting
+	:type ip_based: Boolean
 
 	:returns: a decorator function that limit the number of requests per endpoint
 	"""
 	def ratelimit_decorator(fun):
 		@wraps(fun)
 		def wrapper(*args, **kwargs):
 			# Do not apply rate limits if method is not opted to check
-			if methods != 'ALL' and frappe.request.method.upper() not in methods:
+			if methods != 'ALL' and frappe.request and frappe.request.method and frappe.request.method.upper() not in methods:
```

Comment:
You can use frappe.local.request.method instead of frappe.request.method

Reply:
Doesn't work, tried and tested. It throws AttributeError for request.

Reply:
frappe.local.request exists: https://github.com/frappe/frappe/blob/develop/frappe/app.py#L107